### PR TITLE
chore(core): Move utils to .ts

### PR DESCRIPTION
**Pull request by ibgreen**
For #5589 
- First PR that actually introduces .ts files
- Based on typescript tooling PR #6390 , ideally that PR is landed first.

#### Background
- Move core utils to .ts
#### Change List
- Move core utils to .ts
- Light typing improvements

### CODE REVIEW

File: modules/core/src/utils/flatten.ts [hunk 1]
```diff
@@ -23,14 +23,12 @@
...
+ * @param array The array to flatten.
+ * @param filter= - Optional predicate called on each `value` to
  *   determine if it should be included (pushed onto) the resulting array.
- * @param {Function} map= - Optional transform applied to each array elements.
- * @param {Array} result=[] - Optional array to push value into
- * @return {Array} Returns the new flattened array (new array or `result` if provided)
+ * @return Returns the new flattened array (new array or `result` if provided)
  */
-export function flatten(array, filter = () => true) {
+export function flatten(array: any[], filter = (...args) => true): any[] {
```
Comment (Pessimistress) [inline modules/core/src/utils/flatten.ts:31]: There is only ever one argument passed to `filter`.

File: modules/core/src/utils/iterable-utils.ts [hunk 2]
```diff
@@ -26,8 +26,19 @@ const placeholderArray = [];
...
+export function createIterable(
+  data,
+  startRow = 0,
+  endRow = Infinity
+): {
+  iterable: Iterable<any>;
+  objectInfo: {
+    index: number;
+    data: any;
+    target: any[];
```
Comment (Pessimistress) [inline modules/core/src/utils/iterable-utils.ts:38]: I believe `target` is expected to be `number[]`

File: modules/core/src/utils/positions.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ export function parsePosition(value) {
       };
 
     case 'string':
-      const match = value.match(PERCENT_OR_PIXELS_REGEX);
+      const match = PERCENT_OR_PIXELS_REGEX.exec(value);
```
Comment (Pessimistress) [inline modules/core/src/utils/positions.ts:13]: Why this change? I believe `exec` makes the RegExp object stateful.

File: modules/mesh-layers/package.json [hunk 4]
```diff
@@ -33,6 +33,8 @@
   },
   "dependencies": {
     "@loaders.gl/gltf": "^3.0.8",
+    "@luma.gl/constants": "^8.5.10",
+    "@luma.gl/core": "^8.5.10",
```
Comment (Pessimistress) [inline modules/mesh-layers/package.json:37]: What prompted the dependency change? luma.gl core should be peer dependency even if it's needed. Conflicting copies will break the build.

File: modules/core/src/utils/flatten.ts (hunk 1 above)
Comment (ibgreen) [reply modules/core/src/utils/flatten.ts:31]:
```suggestion
export function flatten(array: any[], filter = (arg: any): boolean => true): any[] {
```

File: modules/core/src/utils/iterable-utils.ts (hunk 2 above)
Comment (ibgreen) [inline modules/core/src/utils/iterable-utils.ts:38]:
```suggestion
    target: number[];
```

File: modules/core/src/utils/positions.ts (hunk 3 above)
Comment (ibgreen) [inline modules/core/src/utils/positions.ts:13]:
```suggestion
    // prettier-ignore
    const match = value.match(PERCENT_OR_PIXELS_REGEX);
```

**Comment by ibgreen** [comment]
Closing in favor of #6452  comments addressed.
